File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/EntityDetailFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.GraphClient
import com.nexussoft.verbum.clients.TimelineClient
import com.nexussoft.verbum.common.arch.Effect
import com.nexussoft.verbum.common.arch.Reducer
import com.nexussoft.verbum.common.arch.only
import com.nexussoft.verbum.common.arch.runEffect
import com.nexussoft.verbum.common.arch.with
import com.nexussoft.verbum.models.BibleEntity
import com.nexussoft.verbum.models.BibleEntityType
import com.nexussoft.verbum.models.EntityDetail
import com.nexussoft.verbum.models.EntityId
import com.nexussoft.verbum.models.GraphSnapshot
import com.nexussoft.verbum.models.PassageReference
import com.nexussoft.verbum.models.SourceReference
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/** One entity's page (docs/PRODUCT.md §9). Twin of the iOS `EntityDetailFeature`. */
object EntityDetailFeature {
    data class State(val entityId: EntityId, val content: Content = Content.Idle)

    sealed interface Content {
        data object Idle : Content
        data object Loading : Content
        data class Loaded(val page: Page) : Content
        data object Failed : Content
    }

    /** [isOnTimeline]: whether the timeline has this entity (§8.2 "Open in timeline"). */
    data class Page(val detail: EntityDetail, val neighborhood: GraphSnapshot, val isOnTimeline: Boolean = false) {
        val entity: BibleEntity get() = detail.entity
        fun related(type: BibleEntityType): List<BibleEntity> = neighborhood.nodes(type)

        val passages: List<PassageReference>
            get() = detail.keyPassages.ifEmpty { neighborhood.nodes(BibleEntityType.PASSAGE).mapNotNull(::passageReference) }

        val sources: List<SourceReference>
            get() {
                val fromEdges = neighborhood.edges.flatMap { it.sourceReferenceIds }.mapNotNull { id -> detail.sources.firstOrNull { it.id == id } }
                return (detail.sources + fromEdges).distinctBy { it.id }
            }
    }

    sealed interface Action {
        data object Started : Action
        data object RetryTapped : Action
        data class PageLoaded(val page: Page) : Action
        data object PageFailed : Action
        data class PassageTapped(val reference: PassageReference) : Action
        data class EntityTapped(val entity: BibleEntity) : Action
        data object GraphTapped : Action
        data object TimelineTapped : Action
        data class Delegate(val delegate: DelegateAction) : Action
    }

    sealed interface DelegateAction {
        data class OpenPassage(val reference: PassageReference) : DelegateAction
        data class OpenEntity(val entity: BibleEntity) : DelegateAction
        /** §8: the graph starts from the selected entity. */
        data class OpenGraph(val entityId: EntityId) : DelegateAction
        /** The timeline, scrolled to this entity's events. */
        data class OpenTimeline(val entityId: EntityId) : DelegateAction
    }

    /** §8.1: default depth one, at most ~8–12 visible nodes. */
    const val NEIGHBOR_LIMIT = 12

    private object LoadId

    /** [timeline] is optional context: if it fails, the page still loads. */
    fun reducer(graphClient: GraphClient, timeline: TimelineClient? = null): Reducer<State, Action> = Reducer { state, action ->
        when (action) {
            Action.Started, Action.RetryTapped -> state.copy(content = Content.Loading).with(
                runEffect(id = LoadId, cancelInFlight = true) { send ->
                    try {
                        val page = coroutineScope {
                            val detail = async { graphClient.detail(state.entityId) }
                            val neighborhood = async { graphClient.neighbors(state.entityId, NEIGHBOR_LIMIT) }
                            val onTimeline = async { timeline?.let { t -> runCatching { t.eventsFor(state.entityId).isNotEmpty() }.getOrDefault(false) } ?: false }
                            Page(detail.await(), neighborhood.await(), onTimeline.await())
                        }
                        send(Action.PageLoaded(page))
                    } catch (e: CancellationException) {
                        throw e
                    } catch (_: Exception) {
                        send(Action.PageFailed)
                    }
                },
            )
            is Action.PageLoaded -> state.copy(content = Content.Loaded(action.page)).only()
            Action.PageFailed -> state.copy(content = Content.Failed).only()
            is Action.PassageTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenPassage(action.reference))))
            is Action.EntityTapped -> {
                val reference = passageReference(action.entity)
                if (reference != null) state.with(Effect.Send(Action.Delegate(DelegateAction.OpenPassage(reference))))
                else state.with(Effect.Send(Action.Delegate(DelegateAction.OpenEntity(action.entity))))
            }
            Action.GraphTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenGraph(state.entityId))))
            Action.TimelineTapped -> state.with(Effect.Send(Action.Delegate(DelegateAction.OpenTimeline(state.entityId))))
            is Action.Delegate -> state.only()
        }
    }

    /** Passage nodes carry their reference in the id (`passage.1Sam.17`). */
    internal fun passageReference(node: BibleEntity): PassageReference? {
        if (node.type != BibleEntityType.PASSAGE) return null
        val parts = node.id.split(".")
        if (parts.size != 3 || parts[0] != "passage") return null
        return PassageReference(parts[1], parts[2].toIntOrNull() ?: return null)
    }
}

File: VerbumKit/Sources/Features/EntityDetail/EntityDetailFeature.swift
import Clients
import ComposableArchitecture
import Foundation
import Models

/// One entity's page (spec §9): the facts, the passages that matter, and the
/// one-hop neighbourhood grouped by kind. Passages go to the reader; related
/// entities push another page. Sources are always listed (§33).
@Reducer
public struct EntityDetailFeature {
    @ObservableState
    public struct State: Equatable, Identifiable {
        public let entityID: EntityID
        public var content: Content = .idle

        public init(entityID: EntityID) {
            self.entityID = entityID
        }

        public var id: EntityID { entityID }
    }

    public enum Content: Equatable, Sendable {
        case idle
        case loading
        case loaded(Page)
        case failed
    }

    /// Everything the view needs, assembled once from the two client calls.
    public struct Page: Equatable, Sendable {
        public let detail: EntityDetail
        public let neighborhood: GraphSnapshot
        /// Whether the timeline has this entity (spec §8.2 "Open in timeline").
        public let isOnTimeline: Bool

        public init(detail: EntityDetail, neighborhood: GraphSnapshot, isOnTimeline: Bool = false) {
            self.detail = detail
            self.neighborhood = neighborhood
            self.isOnTimeline = isOnTimeline
        }

        public var entity: BibleEntity { detail.entity }
        public func related(_ type: BibleEntityType) -> [BibleEntity] { neighborhood.nodes(of: type) }

        /// Detail's key passages, or the passage nodes of the neighbourhood when the detail lists none.
        public var passages: [PassageReference] {
            if !detail.keyPassages.isEmpty { return detail.keyPassages }
            return neighborhood.nodes(of: .passage).compactMap { EntityFixturePassages.reference(for: $0) }
        }

        /// Sources cited by the detail and by every visible edge, without duplicates.
        public var sources: [SourceReference] {
            var seen = Set<String>()
            let fromEdges = neighborhood.edges.flatMap(\.sourceReferenceIds)
            return (detail.sources + fromEdges.compactMap { id in detail.sources.first { $0.id == id } })
                .filter { seen.insert($0.id).inserted }
        }
    }

    public enum Action: Equatable {
        case task
        case retryTapped
        case pageResponse(Result<Page, Failure>)
        case passageTapped(PassageReference)
        case entityTapped(BibleEntity)
        case graphTapped
        case timelineTapped
        case talkTapped
        case delegate(Delegate)

        @CasePathable
        public enum Delegate: Equatable {
            case openPassage(PassageReference)
            case openEntity(BibleEntity)
            /// Spec §8: the graph starts from the selected entity.
            case openGraph(EntityID)
            /// The timeline, scrolled to this entity's events.
            case openTimeline(EntityID)
            /// A spoken conversation about this page.
            case talk(EntityDetail)
        }
    }

    public struct Failure: Error, Equatable, Sendable {
        public init() {}
    }

    @Dependency(\.graphClient) var graphClient
    @Dependency(\.timelineClient) var timelineClient

    public init() {}

    /// Spec §8.1: default depth one, at most ~8–12 visible nodes.
    static let neighborLimit = 12

    public var body: some ReducerOf<Self> {
        Reduce { state, action in
            switch action {
            case .task, .retryTapped:
                state.content = .loading
                let id = state.entityID
                return .run { [graphClient, timelineClient] send in
                    do {
                        async let detail = graphClient.detail(id: id)
                        async let neighborhood = graphClient.neighbors(id: id, limit: Self.neighborLimit)
                        // The timeline is optional context: if it fails, the page still loads.
                        async let onTimeline = (try? timelineClient.eventsFor(entityID: id))?.isEmpty == false
                        await send(.pageResponse(.success(Page(detail: try detail, neighborhood: try neighborhood, isOnTimeline: await onTimeline))))
                    } catch {
                        await send(.pageResponse(.failure(Failure())))
                    }
                }
                .cancellable(id: CancelID.load, cancelInFlight: true)

            case .pageResponse(.success(let page)):
                state.content = .loaded(page)
                return .none

            case .pageResponse(.failure):
                state.content = .failed
                return .none

            case .passageTapped(let reference):
                return .send(.delegate(.openPassage(reference)))

            case .entityTapped(let entity):
                if let reference = EntityFixturePassages.reference(for: entity) {
                    return .send(.delegate(.openPassage(reference)))
                }
                return .send(.delegate(.openEntity(entity)))

            case .graphTapped:
                return .send(.delegate(.openGraph(state.entityID)))

            case .timelineTapped:
                return .send(.delegate(.openTimeline(state.entityID)))

            case .talkTapped:
                guard case .loaded(let page) = state.content else { return .none }
                return .send(.delegate(.talk(page.detail)))

            case .delegate:
                return .none
            }
        }
    }

    private enum CancelID { case load }
}

/// Passage nodes carry their reference in the id (`passage.1Sam.17`). Kept
/// next to the feature so the view never parses ids.
enum EntityFixturePassages {
    static func reference(for node: BibleEntity) -> PassageReference? {
        guard node.type == .passage else { return nil }
        let parts = node.id.split(separator: ".")
        guard parts.count == 3, parts[0] == "passage", let chapter = Int(parts[2]) else { return nil }
        return PassageReference(bookId: String(parts[1]), chapter: chapter)
    }
}
